src/WebJobs.Script/Host/HostIdValidator.cs

```diff
@@ -0,0 +1,198 @@
+
+                var primaryHostStateProvider = _serviceProvider.GetScriptHostServiceOrNull<IPrimaryHostStateProvider>();
+                if (primaryHostStateProvider != null && !primaryHostStateProvider.IsPrimary)
+                {
+                    // as an optimization, only validate on the primary host
+                    return;
+                }
+
+                HostIdInfo hostIdInfo = await ReadHostIdInfoAsync(hostId);
+
+                if (hostIdInfo != null)
+                {
+                    // an existing record exists for this host ID
+                    // verify the host name is the same as our host name
+                    if (string.Compare(_hostNameProvider.Value, hostIdInfo.Hostname, StringComparison.OrdinalIgnoreCase) != 0)
+                    {
+                        HandleCollision(hostId);
+                    }
+                }
+                else
+                {
+                    // no existing record, so write one, claiming this host ID for this host name
+                    // in this storage account
+                    hostIdInfo = new HostIdInfo
+                    {
+                        Hostname = _hostNameProvider.Value
+                    };
+                    await WriteHostIdAsync(hostId, hostIdInfo);
+                }
+            }
+            catch (Exception ex)
+            {
+                // best effort - log error and continue
+                _logger.LogError(ex, "Error validating host ID usage.");
+            }
+        }
+
+        private void HandleCollision(string hostId)
+        {
+            // see if the user has specified a level, otherwise default
+            LogLevel level = DefaultLevel;
+            string value = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsHostIdCheckLevel);
+            if (value != null && Enum.TryParse<LogLevel>(value, out LogLevel parsedLevel))
+            {
+                level = parsedLevel;
+            }
+
+            string message = string.Format(Resources.HostIdCollisionFormat, hostId);
+            if (level == LogLevel.Error)
+            {
+                _logger.LogError(message);
+                _applicationLifetime.StopApplication();
+            }
+            else
+            {
+                // we only allow Warning/Error levels to be specified, so anything other than
+                // Error is treated as warning
+                _logger.LogWarning(message);
+            }
+        }
+
+        internal async Task WriteHostIdAsync(string hostId, HostIdInfo hostIdInfo)
+        {
+            try
+            {
+                var containerClient = _storageProvider.GetBlobContainerClient();
+                string blobPath = string.Format(BlobPathFormat, hostId);
+                BlobClient blobClient = containerClient.GetBlobClient(blobPath);
+                BinaryData data = BinaryData.FromObjectAsJson(hostIdInfo);
+                await blobClient.UploadAsync(data);
```

Comment:
This overload will overwrite an existing blob if one exists, so that would cause a problem if another host had "claimed" the blob/storage. We should update this to use one of the overloads that throws if the blob exists (this would occur if there's a race condition and the blob was created by another app between the check and this write operation)

Reply:
No, the default for overwrite is false. This update will fail if the blob already exists.

Reply:
Are you sure? According to the docs, that isn't the case with this overload (it's inconsistent with some others): https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobclient.uploadasync?view=azure-dotnet#Azure_Storage_Blobs_BlobClient_UploadAsync_System_BinaryData_

It's possible the docs are wrong, it looks like this will overwrite an existing blob.

Reply:
Yes I'm sure. It results in a 409. Please give it a try.

Reply:
I believe you :) Looks like just bad docs.